tscat_gui/utils/keyword_list.py

```diff
@@ -0,0 +1,172 @@
+from .flow_layout import FlowLayout
+
+from PySide2 import QtWidgets, QtCore, QtGui
+
+# UTF-8 letters in the begging, then also numbers and underscore
+_tag_validation_regex = QtCore.QRegularExpression(r'\p{L}\w*')
+
+
+class _Keyword(QtWidgets.QFrame):
+    updated = QtCore.Signal(QtWidgets.QWidget)
+    closed = QtCore.Signal(QtWidgets.QWidget)
+
+    class Label(QtWidgets.QLabel):
+        clicked = QtCore.Signal()
+
+        def __init__(self, text: str, parent=None):
+            super().__init__(text, parent)
+
+        def mousePressEvent(self, event: QtGui.QMouseEvent) -> None:
+            if event.button() == QtCore.Qt.MouseButton.LeftButton:
+                self.clicked.emit()
+            super().mousePressEvent(event)
+
+    class DeleteLabel(QtWidgets.QLabel):
+        clicked = QtCore.Signal()
+
+        def __init__(self, parent=None):
+            super().__init__('✕', parent)
+            self.setMargin(1)
+            self.setObjectName('DeleteLabel')
+            self.setStyleSheet(
+                "#DeleteLabel { border-radius: 5px; background: #444; min-width: 10px; min-height: 10px; }")
+
+        def enterEvent(self, event: QtCore.QEvent) -> None:
+            self.setText('✖')
+            super().enterEvent(event)
+
+        def leaveEvent(self, event: QtCore.QEvent) -> None:
+            self.setText('✕')
+            super().leaveEvent(event)
+
+        def mousePressEvent(self, event: QtGui.QMouseEvent) -> None:
+            if event.button() == QtCore.Qt.MouseButton.LeftButton:
+                self.clicked.emit()
+
+            super().mousePressEvent(event)
+
+    def __init__(self, text: str, tag_list: QtCore.QObject, parent=None):
+        super().__init__(parent)
+
+        self.text = text
+        self.tag_list = tag_list
+
+        self.setObjectName('OneKeyword')
+        self.setStyleSheet("#OneKeyword { border-radius: 15px; background: #999; }")
+
+        self.delete_label = _Keyword.DeleteLabel()
+        self.delete_label.clicked.connect(lambda: self.closed.emit(self))
+
+        self.layout = QtWidgets.QHBoxLayout()
+        self.layout.addWidget(self.delete_label)
+        self.layout.setMargin(5)
```

Comment:
All those margin = 5 aren't going to do weird stuff on different screen densities? For example between a 24"@3k screen and a 14"@4K

Reply:
Good point, Width and height of label scale with font-sizes and their DPI. I don't know how to reflect that here with margin size. I think here I only wanted some distance between the text-label/edit and the delete-button. If it is tiny on a small hi-res, hi-DPI screen, it might not be a problem.